### PR TITLE
nuttx/arch: remove the custom board check in up_testset implementation

### PR DESCRIPTION
**Pull request by extinguish**
## Summary

The `up_testset` implementation is common code, should not add custom board check.
If there are new boards that need to use their own `up_testset()` implementation, for example, `CONFIG_ARCH_CHIP_1XXX`, has provided thier own `up_testset()` implementation, in order to working with `CONFIG_ARCH_CHIP_1XXX`,
we need to change the implemetaton to:
```
#if defined(CONFIG_ARCH_HAVE_TESTSET) \
    && !defined(CONFIG_ARCH_CHIP_LC823450) \
    && !defined(CONFIG_ARCH_CHIP_CXD56XX) \
    && !defined(CONFIG_ARCH_CHIP_RP2040)  \
    && !defined(CONFIG_ARCH_CHIP_1XXX)
static inline_function spinlock_t up_testset(volatile spinlock_t *lock)
{
```

which would make the macro check becoming longer and longer;
beside from this, if the board of `CONFIG_ARCH_CHIP_1XXX` will not upstream to github, then this code will also cannot upstream to github;

beside from the upper cause, another reason for introducing a new configuration：`CONFIG_ARCH_HAVE_CUSTOM_TESTSET` is that generally a board belongs to an architecture. For example, CONFIG_ARCH_CHIP_LC823450 belongs to ARMV7_M. Then ARMV7_M itself has `CONFIG_ARCH_HAVE_TESTSET` enabled.
In this case, all the boards under ARMV7_M have `CONFIG_ARCH_HAVE_TESTSET` enabled by default.
However, `CONFIG_ARCH_CHIP_LC823450` doesn't require `CONFIG_ARCH_HAVE_TESTSET`. We can't turn off `CONFIG_ARCH_HAVE_TESTSET `for all the boards under the entire `ARMV7_M` just for the sake of this single `CONFIG_ARCH_CHIP_LC823450.`
Therefore, we have introduced a new option called `CONFIG_ARCH_HAVE_CUSTOM_TESTSET`.

## Impact

If some boards need custom `up_testset` implementation, should not to add macro control on `arch/arm/include/spinlock.h`, just modify the board config inside `arch/arm/Kconfig` will work.

## Testing

has passed ostest

### CODE REVIEW

**Comment by nuttxpr** [comment]
[**\[Experimental Bot, please feedback here\]**](https://github.com/search?q=repo%3Aapache%2Fnuttx+13552&type=issues)



Yes, this PR appears to meet the NuttX requirements, although the testing section could be improved.

Here's a breakdown:

**Strengths:**

* **Clear Summary:**  The summary explains the problem (increasingly complex macro checks in `up_testset`), the proposed solution (`CONFIG_ARCH_HAVE_CUSTOM_TESTSET`), and the reasoning behind it (avoiding long macros and accommodating non-upstreamed boards).  The explanation regarding architecture-level vs. board-level configuration is helpful.
* **Impact Description:** The impact section clearly states the change for users (modifying board config) and avoids unnecessary changes to the core code.
* **Addresses a Real Problem:** The proposed change tackles maintainability issues arising from conditional compilation.


**Weaknesses:**

* **Testing is Insufficient:**  "has passed ostest" is not enough detail.  You need to provide:
    * Specific build host information (OS, compiler version, etc.)
    * Specific target information (architecture, board, configuration)
    * Actual log output before and after the change, demonstrating the difference in behavior or resource usage.  Even if the output is identical, showing the logs proves that the tests still pass with the new configuration.


**Recommendations to improve the PR:**

1. **Expand the Testing Section:** Provide the specific details mentioned above.  Show the commands used to run the tests.  If `ostest` is a suite of tests, list the specific tests within that suite that were run.  If there's a noticeable improvement (e.g., reduced code size, faster execution), quantify it.
2. **Consider a Before/After Code Example:** Include a small snippet showing how the `Kconfig` would be modified for a board using a custom `up_testset`.  This would enhance clarity for reviewers.


By addressing these points, you'll make your PR stronger and easier for reviewers to assess.

File: arch/arm/Kconfig [hunk 1]
```diff
@@ -344,11 +345,11 @@ config ARCH_CHIP_RP2040
 	select ARCH_CORTEXM0
 	select ARCH_HAVE_RAMVECTORS
 	select ARCH_HAVE_MULTICPU
-	select ARCH_HAVE_TESTSET
 	select ARCH_HAVE_I2CRESET
 	select ARM_HAVE_WFE_SEV
 	select ARCH_HAVE_PWM_MULTICHAN
 	select ARCH_BOARD_COMMON
+	select ARCH_HAVE_CUSTOM_TESTSET
```
Comment (hujun260) [inline arch/arm/Kconfig:352]: Add ARCH_HAVE_CUSTOM_TESTSET to the deconfig of the corresponding architecture.

Comment (extinguish) [reply arch/arm/Kconfig:352]: > Add ARCH_HAVE_CUSTOM_TESTSET to the deconfig of the corresponding architecture.

in NuttX, there are more than 100 defconfigs using `CONFIG_ARCH_CHIP_RP2040` in total. If we were to modify the defconfigs, we could have to modify over 100 defconfig files. for example the following boards defconfig all need to modify:
```
boards/arm/rp2040/raspberrypi-pico-w/configs/telnet/defconfig:19:CONFIG_ARCH_CHIP_RP2040=y
boards/arm/rp2040/raspberrypi-pico-w/configs/smp/defconfig:18:CONFIG_ARCH_CHIP_RP2040=y
boards/arm/rp2040/raspberrypi-pico-w/configs/usbmsc/defconfig:21:CONFIG_ARCH_CHIP_RP2040=y
boards/arm/rp2040/raspberrypi-pico-w/configs/audiopack/defconfig:25:CONFIG_ARCH_CHIP_RP2040=y
boards/arm/rp2040/raspberrypi-pico-w/configs/ssd1306/defconfig:21:CONFIG_ARCH_CHIP_RP2040=y
boards/arm/rp2040/raspberrypi-pico-w/configs/composite/defconfig:21:CONFIG_ARCH_CHIP_RP2040=y
boards/arm/rp2040/raspberrypi-pico-w/configs/displaypack/defconfig:24:CONFIG_ARCH_CHIP_RP2040=y
boards/arm/rp2040/raspberrypi-pico-w/configs/waveshare-lcd-1.3/defconfig:23:CONFIG_ARCH_CHIP_RP2040=y
boards/arm/rp2040/raspberrypi-pico-w/configs/st7735/defconfig:18:CONFIG_ARCH_CHIP_RP2040=y
boards/arm/rp2040/raspberrypi-pico-w/configs/nsh-flash/defconfig:17:CONFIG_ARCH_CHIP_RP2040=y
boards/arm/rp2040/raspberrypi-pico-w/configs/lcd1602/defconfig:18:CONFIG_ARCH_CHIP_RP2040=y
...
```

File: boards/arm/rp2040/raspberrypi-pico/configs/smp/defconfig [hunk 2]
```diff
@@ -37,18 +37,14 @@ CONFIG_RAM_SIZE=270336
 CONFIG_RAM_START=0x20000000
 CONFIG_READLINE_CMD_HISTORY=y
 CONFIG_RR_INTERVAL=200
-CONFIG_SMP=y
```
Comment (anchao) [inline boards/arm/rp2040/raspberrypi-pico/configs/smp/defconfig:40]: this is smp defconfig, why remove SMP?

Comment (extinguish) [reply boards/arm/rp2040/raspberrypi-pico/configs/smp/defconfig]: > this is smp defconfig, why remove SMP?

is a mistake I made, fixed it with another way, please help take a review again

File: arch/Kconfig [hunk 3]
```diff
@@ -489,6 +489,10 @@ config ARCH_HAVE_TESTSET
 	bool
 	default n
 
+config ARCH_HAVE_CUSTOM_TESTSET
+	bool
+	default n
```
Comment (xiaoxiang781216) [inline arch/Kconfig:494]: let's select ARCH_HAVE_TESTSET

File: include/nuttx/spinlock.h [hunk 4]
```diff
@@ -176,7 +176,7 @@ extern volatile uint8_t g_irq_spin_count[CONFIG_SMP_NCPUS];
  *
  ****************************************************************************/
 
-#if defined(CONFIG_ARCH_HAVE_TESTSET)
+#if defined(CONFIG_ARCH_HAVE_TESTSET) || defined(CONFIG_ARCH_HAVE_CUSTOM_TESTSET)
```
Comment (xiaoxiang781216) [inline include/nuttx/spinlock.h:179]:
```suggestion
#ifndef CONFIG_ARCH_HAVE_TESTSET
static inline spinlock_t up_testset(FAR volatile spinlock_t *lock)
```

File: sched/Kconfig [hunk 5]
```diff
@@ -365,7 +365,7 @@ config SMP
 	bool "Symmetric Multi-Processing (SMP)"
 	default n
 	depends on ARCH_HAVE_MULTICPU
-	depends on ARCH_HAVE_TESTSET
+	depends on ARCH_HAVE_TESTSET || ARCH_HAVE_CUSTOM_TESTSET
```
Comment (xiaoxiang781216) [inline sched/Kconfig:368]: revert

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**